lib/temporal/interceptor/client.rb

```diff
@@ -0,0 +1,90 @@
+module Temporal
+  module Interceptor
+    class Client
```

Comment:
I'm not sure I understand the reason for it. I'm happy to rename this to an `OutboundClient` if it provides more clarity. But what's the practical reason? They don't have any interface overlap and I only see disadvantage in having to create 2 classes to implement a simple interceptor.

Reply:
The primary practical reason is to separate the interceptor init from the interceptor provided.

Another important reason is that it allows people to have a single interceptor for clients and workers in the same class. They only have to (optionally) implement `intercept_client`, `intercept_activity`, and/or `intercept_workflow` (though it's `workflow_interceptor_class` for sandbox reasons). So for example, to have OpenTelemetry work everywhere, all you need is the OpenTelemetry interceptor, not one for client and another for workflow (this is also a great benefit of giving the client to the worker instead of a connection, now you get its interceptors, data converters, etc). Making all of these top level means you have to worry about method name clashing.

Reply:
This is something that the user might to choose to implement if they want to and the current solution doesn't restrict this in any way. But I don't see a reason to force this pattern from the SDK point. Moreover as far as I can tell other SDKs don't have a similar solution (at least TS).

Reply:
As far as I can tell you suggesting also requires you to have 2 classes for each interceptor. So I have a logging interceptor and a tracing interceptors for Client I'll need to implement 4 classes. Or am I missing something?

Reply:
You may have to implement 4 classes internally, but from a user API POV, they get 1. Here, if I want 1 user-facing interceptor class for client and activity and worker, we'll have to make sure names don't clash, but it can work.

Reply:
> You may have to implement 2 classes internally, but from a user API POV, they get 1

The contract for a client interceptor in Python SDK is — you supply an object that implements `intercept_client`, that returns another object that implements interceptor methods (`start_workflow(input)`, etc), right?

But I can implement the same thing in Ruby if I want to:

```ruby
class ClientInterceptor < Temporal::Interceptor::Client
  def start_workflow(input)
    ...
  end
end

class CombinedInterceptor
  def intercept_client
    return ClientInterceptor.new
  end
  ...
end
```

And then just provide `CombinedInterceptor.new.intercept_client` to the Client. Isn't this the same exact thing, we're just not forcing you to do this for simpler cases when you don't need to combine all the interceptors together?

Reply:
Yup, almost the exact same thing. Will have to see what steps you have to jump through to, say, support OpenTelemetry. In Python this is `interceptors=[opentelemetry.TracingInterceptor()]` and everything works for clients and workers. I understand if this level of simplicity isn't a goal though.